qiskit/circuit/quantumcircuit.py

```diff
@@ -1691,72 +1690,51 @@ def qasm(
-                    bit_labels[clbit],
-                )
+                instruction_qasm = f"measure {bit_labels[qubit]} -> {bit_labels[clbit]};"
             elif operation.name == "reset":
-                string_temp += f"reset {bit_labels[instruction.qubits[0]]};\n"
+                instruction_qasm = f"reset {bit_labels[instruction.qubits[0]]};"
             elif operation.name == "barrier":
                 qargs = ",".join(bit_labels[q] for q in instruction.qubits)
-                string_temp += "barrier;\n" if not qargs else f"barrier {qargs};\n"
+                instruction_qasm = "barrier;" if not qargs else f"barrier {qargs};"
             else:
-                # Check instructions names or label are valid
-                escaped = _qasm_escape_name(operation.name, "gate_")
-                if escaped != operation.name:
-                    operation = operation.copy(name=escaped)
-
-                # decompose gate using definitions if they are not defined in OpenQASM2
-                if operation.name not in existing_gate_names:
-                    op_qasm_name = None
-                    if operation.name == "permutation":
-                        op_qasm_name = getattr(operation, "_qasm_name", None)
-                    if op_qasm_name:
-                        operation = operation.copy(name=op_qasm_name)
-
-                    if operation not in existing_composite_circuits:
-                        if operation.name in [
-                            operation.name for operation in existing_composite_circuits
-                        ]:
-                            # append operation id to name of operation copy to make it unique
-                            operation = operation.copy(name=f"{operation.name}_{id(operation)}")
-
-                        existing_composite_circuits.append(operation)
-
-                        # Strictly speaking, the code below does not work for operations that
-                        # do not have the "definition" method but require a complex (recursive)
-                        # "_qasm_definition". Fortunately, right now we do not have any such operations.
-                        if getattr(operation, "definition", None) is not None:
-                            _add_sub_instruction_to_existing_composite_circuits(
-                                operation, existing_gate_names, existing_composite_circuits
-                            )
+                operation = _qasm2_define_custom_operation(
+                    operation, existing_gate_names, gates_to_define
+                )
 
                 # Insert qasm representation of the original instruction
-                string_temp += "{} {};\n".format(
-                    operation.qasm(),
-                    ",".join([bit_labels[j] for j in instruction.qubits + instruction.clbits]),
+                bits_qasm = ",".join(
+                    bit_labels[j] for j in itertools.chain(instruction.qubits, instruction.clbits)
```

Comment:
why not `instruction.qubits + instruction.clbits` as before?

Reply:
Oh, no particular reason - I probably just didn't notice I'd done it during the rewrite.  `itertools.chain` is the normal idiom I personally use, so I just wouldn't have thought about it.